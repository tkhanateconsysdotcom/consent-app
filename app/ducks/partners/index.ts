import { actions, namespace, reducer } from "./partners.slice";
import selectors from "./partners.selectors";

export default { actions, namespace, reducer, selectors };
