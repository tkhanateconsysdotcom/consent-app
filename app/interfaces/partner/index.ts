import {
  Partner,
  PartnerType,
  PartnerPayload,
  Preference,
} from "./partner.interface";

export { Partner, PartnerType, PartnerPayload, Preference };
