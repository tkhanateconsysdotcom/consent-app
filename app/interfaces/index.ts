import { Partner, PartnerType, PartnerPayload, Preference } from "./partner";

export { Partner, PartnerType, PartnerPayload, Preference };
