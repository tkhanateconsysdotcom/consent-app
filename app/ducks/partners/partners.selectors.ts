import { filter, propEq, find, pathOr, pipe } from "ramda";

import { PartnerType } from "interfaces";

import { namespace } from "./partners.slice";

const getAllPartners = pathOr([], [namespace]);

const getPartnerById = (state, id: string) =>
  find(propEq("id", id))(state[namespace]);

const getPartnersToMitigate = (state) =>
  pipe(getAllPartners, filter(propEq("isMitigated", false)))(state);

const getPartnersMitigated = (state) =>
  pipe(getAllPartners, filter(propEq("isMitigated", true)))(state);

const isTransportPartner = (partner) =>
  partner.partnerType === PartnerType.longterm;

const isFoodPartner = (partner) => partner.partnerType === PartnerType.romantic;

const isStreamingPartner = (partner) =>
  partner.partnerType === PartnerType.platonic;

const isElectricityPartner = (partner) =>
  partner.partnerType === PartnerType.onenight;

const isCustomPartner = (partner) => partner.partnerType === PartnerType.custom;

const isOtherPartner = (partner) =>
  partner.partnerType === PartnerType.custom ||
  partner.partnerType === PartnerType.platonic ||
  partner.partnerType === PartnerType.onenight;

const getTransportPartners = pipe(getAllPartners, filter(isTransportPartner));

const getFoodPartners = pipe(getAllPartners, filter(isFoodPartner));

const getStreamingPartners = pipe(getAllPartners, filter(isStreamingPartner));

const getElectricityPartners = pipe(
  getAllPartners,
  filter(isElectricityPartner)
);

const getCustomPartners = pipe(getAllPartners, filter(isCustomPartner));

const getOtherPartners = pipe(getAllPartners, filter(isOtherPartner));

export default {
  getAllPartners,
  getTransportPartners,
  getFoodPartners,
  getStreamingPartners,
  getElectricityPartners,
  getCustomPartners,
  getOtherPartners,
  getPartnerById,
  getPartnersToMitigate,
  getPartnersMitigated,
};
