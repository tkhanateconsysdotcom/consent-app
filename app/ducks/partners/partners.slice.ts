import { createSlice, PayloadAction } from "@reduxjs/toolkit";

import { Partner } from "interfaces";

const initialState: Partner[] = [];

const partners = createSlice({
  name: "partners",
  initialState,
  reducers: {
    createPartner(state, action: PayloadAction<Partner>) {
      state.push(action.payload);
    },
    deletePartnerById(state, action: PayloadAction<string>) {
      return (state = state.filter((item) => item.id !== action.payload));
    },
  },
});

const { createPartner, deletePartnerById } = partners.actions;

export const actions = { createPartner, deletePartnerById };

export const namespace = partners.name;

export const reducer = partners.reducer;
