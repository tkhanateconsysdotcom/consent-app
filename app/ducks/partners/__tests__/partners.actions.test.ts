import { TransportEnum } from "carbon-footprint";

import { Partner, PartnerType } from "interfaces";

import partners from "../";

describe("Partners actions should", () => {
  it("export expected actions", () =>
    expect(partners.actions).toMatchSnapshot());

  it("be able to create an partner", () => {
    const partner: Partner = {
      preferences: [],
      id: "123",
      creationDate: "2020-01-26T11:04:55.334Z",
      partnerType: PartnerType.custom,
      partnerModelType: TransportEnum.boat,
      value: 200,
      isMitigated: false,
    };

    const expectedAction = {
      type: partners.actions.createPartner.toString(),
      payload: partner,
    };

    expect(partners.actions.createPartner(partner)).toEqual(expectedAction);
  });

  it("be able to delete an partner", () => {
    const expectedAction = {
      type: partners.actions.deletePartnerById.toString(),
      payload: "123",
    };

    expect(partners.actions.deletePartnerById("123")).toEqual(expectedAction);
  });
});
