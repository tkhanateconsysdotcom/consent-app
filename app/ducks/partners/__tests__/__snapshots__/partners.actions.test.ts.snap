// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Partners actions should export expected actions 1`] = `
Object {
  "createPartner": [Function],
  "deletePartnerById": [Function],
}
`;
